app/services/listings/create.rb

```diff
@@ -0,0 +1,50 @@
+module Listings
+  # This service wraps listing purchses in a transaction which verifies that
+  # the purchaser has enough credits to complete this process.
+  class Create
+    # @param listing [Listing] an initialized but not yet persisted listing
+    # @param purchaser [User] the user attempting to purchase the listing
+    # @param cost [Integer] the number of credits the user has to spend
+    # @return [Listings::Create] the service itself, used for the success? check
+    def self.call(listing, purchaser:, cost:)
+      new(listing, purchaser, cost).call
+    end
+
+    def initialize(listing, purchaser, cost)
+      @listing = listing
+      @purchaser = purchaser
+      @cost = cost
+      @success = false
+    end
+
+    def call
+      ActiveRecord::Base.transaction do
+        # Subtract credits
+        enough_credits = Credits::Buyer.call(
+          purchaser: @purchaser,
+          purchase: @listing,
+          cost: @cost,
+        )
+
+        raise ActiveRecord::Rollback unless enough_credits
```

Comment:
This line is new. Without my spec for `"fails if the purchaser doesn't have enough credits"` always was red. Looking at `Credits::Buyer` and the accompanying spec this is understandable, this service doesn't raise, it just returns `false` which won't be enough to abort this transaction.